feat: Support sample weight for MLP

zshu115x opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
sample_weight support will contributes to class_weight support, so this PR contributes to the issue https://github.com/scikit-learn/scikit-learn/issues/9113, and the stalled PR https://github.com/scikit-learn/scikit-learn/pull/11723 

#### What does this implement/fix? Explain your changes.

- [x] Add sample_weight support for MLP

 

   - [x] Add sample_weight support for multiple loss functions:
     - [x] log_loss
     - [x] binary_log_loss
     - [x]  squared_loss
   - [x]  Add sample_weight parameter to fit and partial_fit
     - [x]  Propagate to relevant functions
     - [x]  Apply sample_weight to gradient calculation
     - [x]  Validate sample_weight
     - [x]  Handle zero weights
   - [x]  Add tests
     - [x]  Test directly loss functions sample_weight invariance
     - [x]  Test fit validation of sample_weight
     - [x]  Test sample weight effect for both MLPClassifier and MLPRegressor
     - [x] Test sample weight together with early stopping


#### Any other comments? 
I have a draft PR https://github.com/scikit-learn/scikit-learn/pull/25326 to support class weight for MLP which will depend on this PR for sample weight support first. 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

zach-shu committed:
feat: enable sample weight and class weight for MLPClassifier

zach-shu committed:
test: add unit tests for sample weight

zach-shu committed:
test: add unit test for sample weight

zach-shu committed:
test: fix unit tests with sample weight and class weight

zach-shu committed:
fix: fix the bug with zero sample weights

zach-shu committed:
fix: make class_weight a parameter for fit() instead

zach-shu committed:
fix: fix lint compliants

zach-shu committed:
feat: add support for partial_fit

zach-shu committed:
fix: fix lint compliants

zach-shu committed:
test: add an additional test for same as check_class_weight_classifiers

zach-shu committed:
fix: fix docstring compliants

zach-shu committed:
fix: fix a lint compliant

zach-shu committed:
test: simplify a unit test

zach-shu committed:
fix: fix a lint compliant

zach-shu committed:
fix: remove unnecessary code in unit test

zach-shu committed:
fix: update unit test

zach-shu committed:
fix: Add sample weight support for MLP

zshu115x commented:
Hi @jnothman @dorcoh @thomasjpfan, moving the comment https://github.com/scikit-learn/scikit-learn/pull/11723#issuecomment-1444275388 here... Could you please review this PR to add sample weight support for MLP as the prerequisite for adding class weight support?

(I am working on another PR to support class weight https://github.com/scikit-learn/scikit-learn/pull/25326)

thomasjpfan commented on sklearn/neural_network/_base.py:
```diff
@@ -166,15 +166,23 @@ def squared_loss(y_true, y_pred):
+    if sample_weight is not None:
+        loss = loss * sample_weight
+
+    return loss.mean() / 2
```
When `sample_weight=None`, the operation becomes:

```python
loss = ((y_true - y_pred) ** 2).sum(axis=-1)
return loss.mean() / 2
```

I do not think that is the same as `((y_true - y_pred) ** 2).mean() / 2`

thomasjpfan commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -582,9 +621,10 @@ def _fit_stochastic(
             should_stratify = is_classifier(self) and self.n_outputs_ == 1
             stratify = y if should_stratify else None
-            X, X_val, y, y_val = train_test_split(
+            X, X_val, y, y_val, sample_weight, _ = train_test_split(
```
I think the validation scores also need to be weighted.

We also need to test for `early_stopping=True` when there are sample weights.

thomasjpfan commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -1138,7 +1192,18 @@ def _validate_input(self, X, y, incremental, reset):
+        nonzero_indices = np.nonzero(sample_weight)[0]
+        if (
+            nonzero_indices.shape[0] < sample_weight.shape[0]
+        ):  # filter out zero-weighted samples
```
I can see this is required to get the common test to pass, but I prefer to get the work correctly with zero weight samples. All the other estimators that have `sample_weights` pass the common tests without actively remove samples.

thomasjpfan commented on sklearn/neural_network/tests/test_base.py:
```diff
@@ -26,5 +27,52 @@ def test_binary_log_loss_1_prob_finite():
     assert np.isfinite(loss)
+
+
+def test_loss_functions():
```
I'll prefer separate test for each loss function and test cases where 

- sample weight is zero, i.e. `sample_weight=[0, 1, 1, 0]`
- sample weight is different, i.e. `sample_weight=[1, 2, 3]`

thomasjpfan reviewed (commented):
Thanks for the PR! I left a review, but I do not have the bandwidth to quickly follow up.

thomasjpfan commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -966,3 +967,47 @@ def test_mlp_partial_fit_after_fit(MLPEstimator):
+
+
+@pytest.mark.parametrize("weighted_class", [i for i in range(3)])
+def test_mlp_classifier_with_sample_weights(weighted_class):
```
Also the regressor also needs to be tested.

For the classifier, there needs to be a test with two classes so that `binary_log_loss` is tested with sample weights.

thomasjpfan commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -1138,7 +1192,18 @@ def _validate_input(self, X, y, incremental, reset):
-        return X, y
+
+        # check sample weight
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
```
It's good to have a benchmark with a big enough dataset to check for performance regressions in terms of runtime + memory.

zach-shu committed:
fix: fix loss functions and add more unit tests

zach-shu committed:
fix: apply sample weight to validation data

zach-shu committed:
chore: update doc

zach-shu committed:
test: update a unit test

zach-shu committed:
fix: fix a variable reference issue

zach-shu committed:
fix: fix sample_weight implementation

zach-shu committed:
fix: raise an expcetion when sample_weight is all zeros

zach-shu committed:
test: fix unit tests for sample weights

zach-shu committed:
fix: fix a lint compliant

zach-shu committed:
test: add additional tests for code coverage

zach-shu committed:
fix: revert an unnecessary change

EwoutH commented:
I would really love to have this feature in scikit learn!

adrinjalali commented:
@scikit-learn/core-devs what do we think here? It's in a nice state, do we want to get it to the finish line or close?

lorentzenchr commented:
> do we want to get it to the finish line

+1

OmarManzoor commented:
I think it might be a good idea to complete this given that considerable work has been put in.

lorentzenchr commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -966,3 +967,112 @@ def test_mlp_partial_fit_after_fit(MLPEstimator):
+
+@pytest.mark.parametrize("weighted_class", [i for i in range(2)])
+@pytest.mark.parametrize("X,y", classification_datasets)
+def test_mlp_classifier_with_sample_weights(weighted_class, X, y):
```
From here on, I would prefer tests like https://github.com/scikit-learn/scikit-learn/blob/5e25db7107a70c8be726d184876f92e4837c4752/sklearn/linear_model/tests/test_ridge.py#L2134.

zach-shu committed:
fix: rebase feature branch and address comments

EwoutH commented:
@zshu115x are you still interested in working on driving this PR home?

zshu115x commented:
> @zshu115x are you still interested in working on driving this PR home?

Hey @EwoutH, yes, I am addressing the review comments and trying to get this PR ready again.

zach-shu committed:
fix: fix sample weight support for neural networks

zach-shu committed:
chore: revert a doc change

zach-shu committed:
fix: address comments

zach-shu committed:
chore: fix a doc format complaint

zshu115x commented:
Hi @lorentzenchr, thanks a lot for your review on this PR. I think I have addressed all your comments. Could you please review it again?

lorentzenchr commented:
@zshu115x Could you "resolve conversation" for all review comments that are addresses?

zshu115x commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -343,6 +361,10 @@ def _backprop(self, X, y, activations, deltas, coef_grads, intercept_grads):
+        if sample_weight is not None:
+            if len(sample_weight.shape) == 1:
+                sample_weight = np.expand_dims(sample_weight, axis=1)
+            deltas[last] *= sample_weight
```
This is where sample_weight is applied to backprop.

zshu115x commented:
Hi @lorentzenchr @glemaitre @thomasjpfan, do you still plan to review this PR?

lorentzenchr commented:
There are still some open old review comments. Can they be marked resolved?

adamjstewart commented:
+1 on adding this feature. I understand the desire not to add any more features to MLP, but this is more for consistency with other estimators that also have this feature, not something unique to MLPs that would add additional maintenance burden.

zshu115x committed:
fix: rebase and fix conflicts

zshu115x committed:
fix: update changelog

zshu115x committed:
fix: fix changelog

zshu115x committed:
fix: fix the score function and the use of validate_data()

zshu115x committed:
fix: uncheck sample_weight_equivalence check for MLP

zshu115x committed:
fix: update xfail_check tags

zshu115x committed:
fix: add sklearn tags to MLPRegressor

zshu115x committed:
fix: remove wrong field

OmarManzoor commented on doc/whats_new/v1.6.rst:
```diff
@@ -26,6 +26,7 @@ Version 1.6
 
 .. _changes_1_6:
 
+
```
I think this change can be reverted.

OmarManzoor commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -270,19 +278,24 @@ def _loss_grad_lbfgs(
             The ith element contains the amount of change used to update the
             intercept parameters of the ith layer in an iteration.
 
+        sample_weigth : ndarray of shape (n_samples,)
```
```suggestion
        sample_weight : ndarray of shape (n_samples,)
```

adamjstewart replied:
Typo

OmarManzoor commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -1158,6 +1243,7 @@ def _validate_input(self, X, y, incremental, reset):
         # This downcast to bool is to prevent upcasting when working with
         # float32 data
         y = self._label_binarizer.transform(y).astype(bool)
+
```
I think we can revert this change as well.
```suggestion
```

adamjstewart commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -440,8 +462,13 @@ def _fit(self, X, y, incremental=False):
+        # Handle sample_weight
+        if sample_weight is not None:
+            sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+            if sum(sample_weight != 0) == 0:
```
Couldn't we just check if the sum = 0? No need for != 0.

zshu115x replied:
Thank you for the comment. There could be a case where the sum is 0 because there are negative weights, but sample_weight is supposed to be non-negative. I will update it together with setting `ensure_non_negative=True` for `_check_sample_weight()`.

OmarManzoor reviewed (commented):
Thank you for the PR @zshu115x

OmarManzoor commented on sklearn/neural_network/tests/test_base.py:
```diff
@@ -27,3 +27,71 @@ def test_log_loss_1_prob_finite(y_true, y_prob):
     assert np.isfinite(loss)
+
+
+def test_squared_loss():
+    y_true = np.array([[0, 1], [0, 0]])
+    y_pred = np.array([[1.0, 1.0], [4.0, 0.0]])
+
+    assert squared_loss(y_true, y_pred) == pytest.approx(2.125)
+
+    sample_weight = np.array([1, 2])
+    assert squared_loss(y_true, y_pred, sample_weight) == pytest.approx(2.75)
+
+
+def test_log_loss():
+    y_true = np.array([[1, 0, 0], [0, 1, 0]])
+    y_prob = np.exp([[-0.2, -2, -2], [-1, -0.1, -1]])
+
+    assert log_loss(y_true, y_prob) == pytest.approx(0.15)
+
+    sample_weight = np.array([1, 4])
+    assert log_loss(y_true, y_prob, sample_weight) == pytest.approx(0.12)
```
Can a short comment be added to specify where the values like 2.125 and 2.75 come from just to make it clear?

zshu115x replied:
Thanks for your review! I was trying to come up with test data that could get nicer numbers, based on a previous comment https://github.com/scikit-learn/scikit-learn/pull/25646#discussion_r1612173142

zshu115x replied:
@OmarManzoor Do you think a comment is still needed here?

OmarManzoor replied:
I think since we are just computing values and checking for them I think this is fine then. Thanks!

glemaitre committed:
revert unecessary change

glemaitre reviewed (commented):
I pushed a small change to remove the diff on the 1.6 changelog.

zshu115x committed:
fix: address comments and fix a few typos and format issues

zshu115x committed:
fix: improve two test cases

OmarManzoor commented on sklearn/neural_network/_base.py:
```diff
@@ -215,17 +225,20 @@ def binary_log_loss(y_true, y_prob):
-        -(xlogy(y_true, y_prob).sum() + xlogy(1 - y_true, 1 - y_prob).sum())
-        / y_prob.shape[0]
-    )
+
+    temp = xlogy(y_true, y_prob) + xlogy(1 - y_true, 1 - y_prob)
+
+    return -np.average(temp, weights=sample_weight, axis=0).sum()
```
```suggestion
    temp = xlogy(y_true, y_prob) + xlogy(1 - y_true, 1 - y_prob)
    return -np.average(temp, weights=sample_weight, axis=0).sum()
```

OmarManzoor commented on sklearn/neural_network/_base.py:
```diff
@@ -197,10 +205,12 @@ def log_loss(y_true, y_prob):
 
-    return -xlogy(y_true, y_prob).sum() / y_prob.shape[0]
+    temp = xlogy(y_true, y_prob)
+
```
```suggestion
```

OmarManzoor reviewed (approved):
Minor suggestions to remove some additional newlines.
Otherwise LGTM. Thanks @zshu115x

zshu115x committed:
fix: remove unnecessary new lines

zshu115x commented:
# Performance Benchmark
The following results are for model training performance comparisons, specifically for `model.fit(X, y)`. The results are based on a dataset with 10k samples and 200 features. 

Before: without this PR
After: with this PR

## Running Time (s)
<img width="303" alt="Screenshot 2024-10-31 at 9 18 28 PM" src="https://github.com/user-attachments/assets/6d0e57ab-936b-4916-82ca-41c00809afc3">

## Peak Allocated Memory (MB)
<img width="305" alt="Screenshot 2024-10-31 at 9 18 43 PM" src="https://github.com/user-attachments/assets/a67b466c-0653-4c5a-bcfd-a9ce6ddc9bc3">

zshu115x commented:
Thanks a lot for your review @OmarManzoor! 

@lorentzenchr would you please review this PR again?

lorentzenchr commented:
Now that I opened #30155, I prefer it over this one. In particular the scaling of the L2 penalty alpha. (And some smaller things).
I should, however, add @zshu115x as author over there.

zshu115x commented:
> Now that I opened #30155, I prefer it over this one. In particular the scaling of the L2 penalty alpha and partially reinventing the common tests for sw. (And some smaller things). I should, however, add @zshu115x as author over there.

Thank you for the comment @lorentzenchr. I will let others comment too, but my understanding of L2 penalty is that it shouldn’t be affected by sample weight. In general, regularizers apply to the model’s weights as a whole, not to each sample individually. 

In terms of which PR to move forward, since this PR already has an approval, and if we think it is ready, I would prefer to moving forward with it. If you have other comments or want to incorporate some changes you had in the other PR, please let me know. I will update this PR accordingly and add you as an author as well.

Any thoughts from anyone else? I would love to get more reviews and approvals and merge this PR soon. It’s been quite a while since the issue was created and since I first opened this PR. I happen to have some more time recently and would like to get this done. Thank you!

OmarManzoor commented:
Based on my own search, I think that @zshu115x is correct in pointing out that `sample_weights` should not be applied to the regularization term. Even considering our own codebase I think we generally follow this principle. Here is an example of the `loss` function defined in the `LinearModelLoss` class
https://github.com/scikit-learn/scikit-learn/blob/8f620fd7a39d1bf1292b09192575d25cddc7888b/sklearn/linear_model/_linear_loss.py#L251-L264

@glemaitre and @adrinjalali Could you kindly provide your opinion?

lorentzenchr commented:
About scaling of L2 penalty terms, see #15657. There is not single "correct" way of doing it. Here in MLP, the objective currently is: `obj = 1/n * sum(loss) + alpha/(2n) ||coef||`. So we already scale L2 by number of samples as in Ridge regression because it is considered to provide a better default.

For sample weights, I strongly want to avoid to end up with: `loss = 1/sum(sw) * sum(sw * loss) + alpha/(2n) ||coef||`. But this PR does exactly that. I instead prefer `loss = 1/sum(sw) * sum(sw * loss) + alpha/(2*sum(sw)) ||coef||` because it has better invariance properties.

lorentzenchr commented:
And before we change many things here, I still prefer #30155, where the work has already been done (while this PR was kind of stale).

adrinjalali commented:
cc @ogrisel @snath-xoc @antoinebaker on the sample weight issue here.

zshu115x commented:
> > For sample weights, I strongly want to avoid to end up with: `loss = 1/sum(sw) * sum(sw * loss) + alpha/(2n) ||coef||`. But this PR does exactly that. I instead prefer `loss = 1/sum(sw) * sum(sw * loss) + alpha/(2*sum(sw)) ||coef||` because it has better invariance properties.
> 
> Yes you are correct I missed focusing on the part where the current loss actually incorporates a division by the number of samples already. @zshu115x since the current loss function is already dividing the regularization by the number of samples, dividing by the sum of the sample weights in the case where we have weights seems better as mentioned above.

Initially, I had `loss = 1/n * sum(sw * loss) + alpha/(2n) ||coef||`. And I updated it to `loss = 1/sum(sw) * sum(sw * loss) + alpha/(2n) ||coef||` according to @lorentzenchr's comment https://github.com/scikit-learn/scikit-learn/pull/25646#discussion_r1612082409. I think both are correct. The difference is whether we normalize the sample weight within each batch. 

In terms of regularization, I don't think it is a good idea to scale it by sample weight.  

For `loss = 1/sum(sw) * sum(sw * loss) + alpha/(2n) ||coef||`, 
* `1/sum(sw) * sum(sw * loss)` is the loss calculated from samples, where `1/sum(sw)` simply normalizes the sample weight
*  `alpha/(2n) ||coef||` is the regularization term calculated from model parameters. `1/(2n)` is a scaling factor for `alpha`, it is static for a given batch size. If we use `1/(2*sum(sw))` as the scaling factor for regularizers, it basically means we change `alpha` based on sample_weight, and its effect will be different on each batch. This should be avoided. 

In the case where sample weights are identical large numbers, for example, [100, 100, 100, ...], `1/sum(sw) * sum(sw * loss)` will be equivalent to using no sample weight, but `1/(2*sum(sw))` will significantly scale down alpha. I don't think this is the right behavior for regularization.
